azurelinuxagent/common/utils/cryptutil.py

```diff
@@ -53,10 +53,21 @@ def gen_transport_cert(self, prv_file, crt_file):
     def get_pubkey_from_prv(self, file_name):
         if not os.path.exists(file_name):
             raise IOError(errno.ENOENT, "File not found", file_name)
-        else:
-            cmd = [self.openssl_cmd, "rsa", "-in", file_name, "-pubout"]
-            pub = shellutil.run_command(cmd, log_error=True)
-            return pub
+
+        # OpenSSL's pkey command may not be available on older versions so try 'rsa' first.
+        try:
+            command = [self.openssl_cmd, "rsa", "-in", file_name, "-pubout"]
+            return shellutil.run_command(command, log_error=False)
+        except shellutil.CommandError as error:
+            if not ("Not an RSA key" in error.stderr or "expecting an rsa key" in error.stderr):
+                logger.error(
+                    "Command: [{0}], return code: [{1}], stdout: [{2}] stderr: [{3}]",
+                    " ".join(command),
+                    error.returncode,
+                    error.stdout,
+                    error.stderr)
+                raise
+            return shellutil.run_command([self.openssl_cmd, "pkey", "-in", file_name, "-pubout"], log_error=True)
```

Comment:
If pkey is may not be available on older versions of openssl, should we handle any errors here?

Reply:
how should we handle them? we have no way to extract the certs so at this point i'm just letting the goal state fail by letting the exception bubble up

Reply:
If we catch and log the error, could some extensions in the GS which don't require the cert still provision successfully? And then only the extensions which require the cert fail

Reply:
We don't know what extensions require those certs. The tenant cert and ssh keys for provisioning always use RSA (at least currently). EC would be used on KeyVault certs, which are typically used with CSE or something like that, and the scripts may not be resilient to a failure not finding the cert (and may have already done partial work). At this point I would fail the goal state rather than swallowing the error. Once we start getting telemetry about pkey we can revisit this.

Reply:
Makes sense!